flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/stream/StreamingFileCommitter.java

```diff
@@ -0,0 +1,175 @@
+		implements OneInputStreamOperator<StreamingFileCommitter.CommitMessage, Void> {
+
+	private static final long serialVersionUID = 1L;
+
+	private final Configuration conf;
+
+	private final List<String> partitionKeys;
+
+	private final TableMetaStoreFactory metaStoreFactory;
+
+	private transient PartitionCommitManager commitManager;
+
+	private transient TaskTracker taskTracker;
+
+	private transient long currentWatermark = Long.MIN_VALUE;
+
+	private transient List<PartitionCommitPolicy> policies;
+
+	public StreamingFileCommitter(
+			List<String> partitionKeys, TableMetaStoreFactory metaStoreFactory, Configuration conf) {
+		this.partitionKeys = partitionKeys;
+		this.metaStoreFactory = metaStoreFactory;
+		this.conf = conf;
+	}
+
+	@Override
+	public void initializeState(StateInitializationContext context) throws Exception {
+		super.initializeState(context);
+		this.commitManager = new PartitionCommitManager(
+				context.isRestored(),
+				context.getOperatorStateStore(),
+				getUserCodeClassloader(),
+				partitionKeys,
+				conf);
+		this.policies = PartitionCommitPolicy.createCommitChain(
+				conf.get(PARTITION_COMMIT_POLICY_TYPE),
+				conf.get(PARTITION_COMMIT_POLICY_SUCCESS_FILE_NAME));
+	}
+
+	@Override
+	public void processElement(StreamRecord<CommitMessage> element) throws Exception {
+		CommitMessage message = element.getValue();
+		for (String partition : message.partitions) {
+			commitManager.addPartition(partition);
+		}
+
+		if (taskTracker == null) {
+			taskTracker = new TaskTracker(message.numberOfTasks);
+		}
+		boolean needCommit = taskTracker.add(message.checkpointId, message.taskId);
+		if (needCommit) {
+			commitPartitions(commitManager.triggerCommit(message.checkpointId));
+		}
+	}
+
+	private void commitPartitions(List<String> partitions) throws Exception {
+		try (TableMetaStoreFactory.TableMetaStore metaStore = metaStoreFactory.createTableMetaStore()) {
+			FileSystem fs = metaStore.getLocationPath().getFileSystem();
+			for (String partition : partitions) {
+				LinkedHashMap<String, String> partSpec = extractPartitionSpecFromPath(new Path(partition));
+				Path path = new Path(metaStore.getLocationPath(), generatePartitionPath(partSpec));
+				for (PartitionCommitPolicy policy : policies) {
+					policy.commit(partSpec, path, fs, metaStore);
+				}
+			}
+		}
+	}
+
+	@Override
+	public void processWatermark(Watermark mark) throws Exception {
+		super.processWatermark(mark);
+		this.currentWatermark = mark.getTimestamp();
+	}
+
+	@Override
+	public void snapshotState(StateSnapshotContext context) throws Exception {
+		super.snapshotState(context);
+		commitManager.snapshotState(context.getCheckpointId(), currentWatermark);
+	}
+
+	/**
+	 * The message sent upstream.
+	 */
+	public static class CommitMessage implements Serializable {
+
+		public long checkpointId;
+		public int taskId;
+		public int numberOfTasks;
+		public List<String> partitions;
```

Comment:
Can we have some comments about which partitions should be in this list? My understanding is it should include partitions for which some files should be committed, right?

Reply:
Yes.